### ISSUE TITLE
Offer @traced macro when macros gain function body replacement capabilities

### ISSUE THREAD

**Issue by stevapple** [apple/swift-distributed-tracing#125]
Typically we trace at function level by wrapping the whole implementation in a span, as demonstrated by the sample.

```swift
func makeDinner() async throws -> Meal {
  try await withSpan("makeDinner") { _ in
    async let veggies = try chopVegetables()
    async let meat = marinateMeat()
    async let oven = preheatOven(temperature: 350)
    // ...
    return try await cook(veggies, meat, oven)
  }
}
```

It makes me think if we can have an `@Traced` macro attached to functions, which handles the span automatically for users. Eventually such macro should be configurable (explicit tracer, span name, extracting context from parameters, etc.), but AFAIK Swift macros seem not having such level of expressiveness yet.

```swift
@Traced
func makeDinner() async throws -> Meal {
  async let veggies = try chopVegetables()
  async let meat = marinateMeat()
  async let oven = preheatOven(temperature: 350)
  // ...
  return try await cook(veggies, meat, oven)
}
```

**Comment by ktoso** [comment]
Yes, this is planned but we're missing macro functionality to pull it off.

We need "function body macros" -- once those ship we'll ship such macro in this library 👍

**Comment by ktoso** [comment]
rdar://109247127

**Comment by porglezomp** [comment]
Since function body macros ship with Swift 6.0 (https://github.com/swiftlang/swift/pull/73765) is this unblocked now?

**Comment by ktoso** [comment]
Additional info: Backed out the traced macro until we can figure out a policy about macros adoption in this package that would not hinder upgrades of the dependency for folks not even using the macro.
